dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java

```diff
@@ -136,13 +132,42 @@ public TrackerImportReport importTracker( TrackerImportParams params )
             }
             else
             {
+
```

Comment:
I find this new code a bit hard to read (I would try to keep this method as tiny and lean as possible).
What about encapsulating the call to the rule engine like so:

```
@Override
    public TrackerImportReport importTracker( TrackerImportParams params )
    {
        params.setUser( trackerUserService.getUser( params.getUserId() ) );

        // Init the Notifier
        ImportNotifier notifier = new ImportNotifier( this.notifier, params );

        // Keeps track of the elapsed time of each Import stage
        TrackerTimingsStats opsTimer = new TrackerTimingsStats();

        notifier.startImport();

        TrackerValidationReport validationReport = null;

        TrackerBundleReport bundleReport;

        try
        {
            //
            // pre-heat
            //
            TrackerBundle trackerBundle = opsTimer.exec( PREHEAT_OPS,
                () -> preheatBundle( params ) );

            Map<TrackerType, Integer> bundleSize = calculatePayloadSize( trackerBundle );

            //
            // preprocess
            //
            opsTimer.execVoid( PREPROCESS_OPS,
                () -> preProcessBundle( trackerBundle ) );

            //
            // validate
            //
            validationReport = opsTimer.exec( VALIDATION_OPS,
                () -> validateBundle( trackerBundle ) );

            notifier.notifyOps( VALIDATION_OPS, opsTimer );

            if ( exitOnError( validationReport, params ) )
            {
                return buildReportAndNotify( validationReport, opsTimer, bundleSize, notifier );
            }
            else
            {
                validationReport = execRuleEngine(opsTimer, trackerBundle, validationReport, notifier);


                if ( exitOnError( validationReport, params ) )
                {
                    return buildReportAndNotify( validationReport, opsTimer, bundleSize, notifier );
                }

                //
                // commit
                //
                if ( TrackerImportStrategy.DELETE == params.getImportStrategy() )
                {
                    bundleReport = opsTimer.exec( COMMIT_OPS, () -> deleteBundle( trackerBundle ) );
                }
                else
                {
                    bundleReport = opsTimer.exec( COMMIT_OPS, () -> commitBundle( trackerBundle ) );
                }

                notifier.notifyOps( COMMIT_OPS, opsTimer );

                TrackerImportReport trackerImportReport = TrackerImportReport.withImportCompleted( TrackerStatus.OK,
                    bundleReport, validationReport,
                    opsTimer.stopTimer(), bundleSize );

                notifier.endImport( trackerImportReport );

                return trackerImportReport;
            }
        }
        catch ( Exception e )
        {
            log.error( "Exception thrown during import.", e );

            TrackerImportReport report = TrackerImportReport.withError( "Exception:" + e.getMessage(),
                validationReport, opsTimer.stopTimer() );

            notifier.endImportWithError( report, e );

            return report;
        }
    }

    private TrackerValidationReport execRuleEngine( TrackerTimingsStats opsTimer, TrackerBundle trackerBundle,
        TrackerValidationReport report, ImportNotifier notifier )
    {
        //
        // rule engine
        //
        opsTimer.execVoid( PROGRAMRULE_OPS,
            () -> runRuleEngine( trackerBundle ) );

        notifier.notifyOps( PROGRAMRULE_OPS, opsTimer );

        //
        // rule engine
        //
        TrackerValidationReport finalValidationReport = report;
        report = opsTimer.exec( VALIDATE_PROGRAMRULE_OPS,
            () -> validateRuleEngine( trackerBundle, finalValidationReport ) );

        notifier.notifyOps( VALIDATE_PROGRAMRULE_OPS, opsTimer );

        return report;
    }

    private TrackerImportReport buildReportAndNotify( TrackerValidationReport validationReport,
        TrackerTimingsStats opsTimer, Map<TrackerType, Integer> bundleSize, ImportNotifier notifier )
    {
        TrackerImportReport trackerImportReport = TrackerImportReport
                .withValidationErrors( validationReport, opsTimer.stopTimer(),
                        bundleSize.values().stream().mapToInt( Integer::intValue ).sum() );

        notifier.endImport( trackerImportReport );

        return trackerImportReport;
    }

    private boolean exitOnError( TrackerValidationReport validationReport, TrackerImportParams params )
    {
        return validationReport.hasErrors() && params.getAtomicMode() == AtomicMode.ALL;
    }
```